Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/Internal/NamespaceSynchronizationConfig.swift

```diff
@@ -219,28 +160,48 @@ internal class NamespaceSynchronization: Sequence {
         do {
             return Set(
                 try self.docsColl.distinct(
-                    fieldName: CoreDocumentSynchronization.Config.CodingKeys.documentId.rawValue,
+                    fieldName: CoreDocumentSynchronization.CodingKeys.documentId.rawValue,
                     filter: [
-                        CoreDocumentSynchronization.Config.CodingKeys.isStale.rawValue: true,
-                        CoreDocumentSynchronization.Config.CodingKeys.namespace.rawValue:
-                            try BSONEncoder().encode(config.namespace)
+                        CoreDocumentSynchronization.CodingKeys.isStale.rawValue: true,
+                        CoreDocumentSynchronization.CodingKeys.namespace.rawValue:
+                            try BSONEncoder().encode(namespace)
                     ]).compactMap({
                         $0 == nil ? nil : HashableBSONValue($0!)
                     })
             )
         } catch {
-            errorListener?.on(error: error, forDocumentId: nil, in: self.config.namespace)
+            errorListener?.on(error: error, forDocumentId: nil, in: self.namespace)
             return Set()
         }
     }
 
     func set(stale: Bool) throws {
         _ = try nsLock.write {
             try docsColl.updateMany(
-                filter: ["namespace": try BSONEncoder().encode(config.namespace)],
+                filter: ["namespace": try BSONEncoder().encode(namespace)],
                 update: ["$set": [
-                    CoreDocumentSynchronization.Config.CodingKeys.isStale.rawValue: true
+                    CoreDocumentSynchronization.CodingKeys.isStale.rawValue: true
                 ] as Document])
         }
     }
+
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
```

Comment:
[major race condition, but might be worth opening a separate ticket] I think you need to take a `nsLock.read` lock over this entire method. It would be pretty major race condition if this document config could change while it was being persisted, ending up with a partially corrupt persisted config. This is going to be tricky without reentrant lock semantics so it might be worth opening a ticket to have our locks have working reentrant semantics